### ISSUE TITLE
Graphs loading very slowly

### ISSUE THREAD

**Issue by AurimasNav** [kiali/kiali#5743]
Hi,
I've just deployed istio and now adding kiali, but the graphs are loading very slowly:
1 namespace with 2 services 2 apps and 2 edges takes 2-3 secs to load,
2 namespaces with 3 services 3 apps and 6 edges takes 5-6 secs to load,
3 namespaces with 3 services 4 apps and 7 edges takes 8-9 secs and so forth, until it reaches a point where it times out and displays an error instead of graph.

My current state:
Kiali version: 1.62.0
Istio version: 1.16.1
Running on AKS

I have only configured prometheus for istio, no grafana, nor jaeger, also no ingress or egress gateways.
![image](https://user-images.githubusercontent.com/7810047/213690097-714b89c1-3336-4681-a8bd-98cc75618ba0.png)

I have checked prometheus/istio/kiali pods - no unusual restarts, no cpu throttling:
![image](https://user-images.githubusercontent.com/7810047/213690301-b381467a-dc01-4ba2-ac74-01ff9ae8d656.png)
![image](https://user-images.githubusercontent.com/7810047/213690466-e2f5f1c9-75a6-4031-96b6-de8176afa68b.png)
![image](https://user-images.githubusercontent.com/7810047/213690828-37ed095e-bdea-42cc-bbc6-6a2a88ef8863.png)

K8s node resources look OK to me:
![image](https://user-images.githubusercontent.com/7810047/213689723-e46307a0-b4d0-47de-bebf-403ee18845b6.png)

metrics in prometheus seem to be loading normally (though I am not sure I am testing it properly):
![image](https://user-images.githubusercontent.com/7810047/213692169-4722b363-113d-4edc-a552-c07b7ae48074.png)

I have no more ideas where to look for problems.

**Comment by AurimasNav** [comment]
It seems it was actually a node resource problem, I've enabled proxy sidecars for several namespaces and those requested almost all available cpu resources on the nodes (100m per sidecar by default), even though there were unused CPU on nodes and no throttling on the kiali metrics I guess it was not able to increase cpu consumption due to everything being "requested" by other containers.

**Comment by jshaughn** [comment]
Hi @AurimasNav , this definitely seems like a resource issue.  Is there a reason you re-opened?

**Comment by AurimasNav** [comment]
I am unable to identify the bottleneck, was hoping for some tips where else to look.

**Comment by jshaughn** [comment]
Usually slow graph generation is due to low resources for Prometheus in the control plane namepace, or poor network connectivity.

**Comment by AurimasNav** [comment]
I've added another node to k8s, moved kiali and prometheus to it with virtually nothing else running there and it had no effect on graph loading speed.
![image](https://user-images.githubusercontent.com/7810047/214835944-20bcb8f9-2353-42b8-b593-b3435326f276.png)
The pods are nowhere near their resource limits:
- kiali ~10% of cpu limits and ~3% of memory limits
- prometheus ~2% cpu and 18% memory limits

While poking pod dashboards in grafana I've noticed that there is 2.5 MB/s receive bandwidth to kiali, while prometheus only has ~10Kb/s transmit bandwidth, is that expected?

**Comment by jshaughn** [comment]
I can't comment about the bandwith question.  I guess we need to better understand where the time is being spent.  Have you tried accessing the Prometheus console directly and performing some queries, or using Grafana dashboards against the same prom instance?  If so, it is performing well?  Is it just the graph that is slow?  Can you generate workload charts reasonably (In Kiali select a workload and in the details view select the Inbound or Outbound merics.  You can visit the Kiali workload in (by default) istio-system and you should see an additional tab for a custom dashboard named `Kiali Internal Metrics`.  There is a chart called `Graph generation duration`.  That would reflect time on the server for generating the graphs.

All I can say is that something environmental seems to be in play, because what your seeing is anomalous.

**Comment by AurimasNav** [comment]
Grafana dashboards from same prometheus are working fine, prometheus is also performing well.
Inbound/Outbound metrics in workloads load almost instantly, however on kiali internal metrics all charts display 'no data available' (I get an error 'could not fetch spans', is it dependent on tracing component?).

**Comment by jshaughn** [comment]
In the screenshot above you see the red bullet on the manifest, left of the message center?  WHen you hover what does it say?  Something just has to be wrong with the set up here, Kiali is not properly communicating with prom.  Check the Kiali logs as well, via the Logs tab in same screenshot.

**Comment by AurimasNav** [comment]
The red bullet is in my original post, will get back with the logs.

**Comment by jshaughn** [comment]
By the way, the way the graph is built is that each requested namespaces needs to be processed on it's own and then the results are "stitched" together.  So the linear progression of time increase as namespaces increase, is sort of expected, although the times are still egregriously long.

**Comment by AurimasNav** [comment]
I've tried to load graphs for all namespaces (35 in total) waited for around 10 minutes before giving up, nothing was loaded and no error displayed, the logs including that period:

```
2023-01-31T06:12:38Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:13:08Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:13:40Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:14:09Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:14:40Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:15:09Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:15:39Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:16:10Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:16:40Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:17:12Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:17:41Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:18:13Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:18:44Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:19:14Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:19:44Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:20:16Z ERR GetAppTraces, Jaeger GRPC client error: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:20:44Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:21:01Z ERR GetAppTraces, Jaeger GRPC client error: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:21:40Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:22:09Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:22:19Z INF Kiali Cache: Updating cache with new token

2023-01-31T06:22:19Z INF [Kiali Cache] Waiting for cluster-scoped cache to sync

2023-01-31T06:22:19Z INF [Kiali Cache] started

2023-01-31T06:23:04Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:23:34Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:24:04Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:24:04Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

2023-01-31T06:24:05Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup tracing.istio-system on 10.0.0.10:53: no such host"

I0131 06:27:59.953130       1 request.go:601] Waited for 1.002950618s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/keda/statefulsets

I0131 06:29:15.438829       1 request.go:601] Waited for 1.002931942s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/credentials-worker/statefulsets

I0131 06:30:31.134397       1 request.go:601] Waited for 1.001958106s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/ingress-azure/statefulsets

I0131 06:31:46.260461       1 request.go:601] Waited for 1.004454948s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/istio-system/statefulsets

I0131 06:33:01.919054       1 request.go:601] Waited for 1.002003726s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/marketplace/replicasets

I0131 06:34:17.243417       1 request.go:601] Waited for 1.003405371s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/tenant-management/statefulsets

I0131 06:35:32.315061       1 request.go:601] Waited for 1.005115438s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/tenant-management/statefulsets

I0131 06:36:48.977885       1 request.go:601] Waited for 1.000204702s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/apps/v1/namespaces/tag-management/statefulsets

I0131 06:38:08.383719       1 request.go:601] Waited for 1.00161125s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/api/v1/namespaces/frontend/pods

I0131 06:39:23.531360       1 request.go:601] Waited for 1.000775762s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/api/v1/namespaces/tag-management/pods
```

**Comment by jshaughn** [comment]
So those client throttling messages don't look very good.  The graph queries prometheus for metrics but then also queries k8s for information in order to decorate the graph with more information.  That information is typically cached but if the cache is having trouble getting built then it wuld have to query directly.  I wonder if the problem is not with prometheus querying but rather issues with getting the k8s information.  Let's try two queries, the first to generate the graph with "decoration" and the second without.

In your browser or from a command line try (this may work as-is only if you are using `anonymous` auth):

This one has the decoration:

`http://YOURHOST:YOURPORT/kiali/api/namespaces/graph?duration=600s&graphType=workload&appenders=deadNode,istio,serviceEntry,sidecarsCheck,workloadEntry,health&namespaces=YOURNAMESPACE`

This one does not:

`http://YOURHOST:YOURPORT/kiali/api/namespaces/graph?duration=600s&graphType=workload&appenders=&namespaces=YOURNAMESPACE`

The second one should only interact with prometheus.

I don't know why you would be seeing that throttling but I would think it would also slow down other aspects of Kiali.  There are a variety of threads on the internet talking about that issue, maybe check out https://stackoverflow.com/questions/71596906/client-side-throttling-response-from-kubernetes-kubectl-command.

**Comment by AurimasNav** [comment]
The first query with 6 namespaces selected took 15970 ms
The second with the same namespaces took 241 ms

I've seen some of the threads regarding this issue, but I don't quite experience the other symptoms like slow kubectl command responses, maybe only kiali hits some threshold limits in my environment as seen per this graph api server requests are ~10 when kiali is not trying to load graphs, and spikes to hundreds when it is:

![image](https://user-images.githubusercontent.com/7810047/215990814-79b60ab2-947e-4539-ad05-e2247fbfc5dc.png)
basically all the fragments above the floor are when kiali was trying to graphs, my question is are those numbers reasonable or is it some kind of anomaly as well?
![image](https://user-images.githubusercontent.com/7810047/215991306-a3b5efbb-9a7b-439e-ba4d-56873812f8a6.png)

**Comment by jshaughn** [comment]
Thanks @AurimasNav, for continuing to be patient as we try to understand what is happening here.

A couple of things.  First, looking at the logs I think you should probably disable tracing if it's not deployed.  This is very probably unrelated but at least removes a variable.  You can [set](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.tracing.enabled) `spec.external_services.tracing.enabled` false.

In the logs we can see a "token refresh" prior to the client throttling messages.  We are wondering if that refresh has something to do with the problem, as a token refresh does affect the k8s cache.  As a test, you can restart the Kiali pod.  There should not be a token refresh for roughly 45 minutes, assuming you have not customized `spec.login_token.expiration_seconds`.  So, for the first 45 minutes of Kiali pod life, do you see the same slowness?

You could also enable more detailed logging by setting `https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.logger.log_level` to `debug` or `trace`.

**Comment by AurimasNav** [comment]
With 1.61 this issue is not present. Thank you for your time!

**Comment by jshaughn** [comment]
@AurimasNav Excellent.  Thank you so much for helping us discover this performance regression.  It will be fixed in v1.64 and also backported to v1.63, because v1.63 will be the default version for Istio 1.17, when installed as an add-on.

**Comment by nrfox** [comment]
> @AurimasNav Excellent. Thank you so much for helping us discover this performance regression.

+1 thank you for reporting and helping diagnose this issue!

**Comment by jshaughn** [comment]
This will also be backported to v1.62, where the regression first appeared.

**Comment by jshaughn** [comment]
Just to finish this discussion, the regression has been fixed and released in versions 1.62.1 and 1.63.2.  It will of course be in the upcoming 1.64 release as well.
